### PR TITLE
Validator stake & delegation table in caching layer. 

### LINKED ISSUE #1026: Validator stake & delegation table in caching layer. Research + Implement.
Research design and implement the way to track delegation share to total share in cacher. Must fully understand lifecycle of validator delegation and staking. Expect the task owner to talk through that clearly and implement respective cacher.

### LINKED ISSUE #1119: Update account balance during EndBlock for mature undelegation
NOTE: There are no transfer events. The accounts are set directly! See code in `cosmos-sdk/x/staking/handler.go`'s `EndBlocker` function. Mature unbonding tokens will be sent to users. Please verify that your code works correctly by delegating and unbonding.

### PR DESCRIPTION
**Pull request by taobun**
fixed #1026 
fixed #1119

- [x] Create validator & delegate in genesis
- [x] Handle create validator msg
- [x] Handle delegate msg
- [x] Handle redelegate msg
- [x] Handle undelegate(unbond) msg
- [x] Update account balance after unbonding period correctly

### CODE REVIEW

File: chain/db/endblock.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+					panic(err)
+				}
+				rawResult = result.Data
+			}
+			err = b.ResolveRequest(requestID, resolveStatus, rawResult)
+			if err != nil {
+				panic(err)
+			}
+		}
+	case staking.EventTypeCompleteUnbonding:
```
Comment (sorawit) [inline chain/db/endblock.go:53]: So I think this fixes #1119?

File: chain/db/delegation.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	delegatorAddress sdk.AccAddress,
+	validatorAddress sdk.ValAddress,
+) error {
+	return b.tx.Delete(Delegation{
+		DelegatorAddress: delegatorAddress.String(),
+		ValidatorAddress: validatorAddress.String(),
+	}).Error
+}
+
+func (b *BandDB) delegate(
```
Comment (sorawit) [inline chain/db/delegation.go:32]: Can a validator delegate for himself? Does `SelfDelegation` need to increase?

File: chain/db/slashing.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package db
```
Comment (sorawit) [inline chain/db/slashing.go:1]: Let's handle the case when a validator got jailed too. Jail should become true in that case?

**Review by sorawit — CHANGES_REQUESTED** [review]
How do we compute validator stake? I the PR say "Validator stake", but its't this PR only for delegation share?

File: chain/db/delegation.go (hunk 2 above)
Comment (taobun) [reply chain/db/delegation.go:32]: SelfDelegation is just a delegation record, not a variable in validator struct, logic query delegation by validator to compare with MinSelfDelegation

File: chain/db/endblock.go (hunk 1 above)
Comment (taobun) [reply chain/db/endblock.go:53]: yep

File: chain/db/delegation.go (hunk 2 above)
Comment (sorawit) [reply chain/db/delegation.go:32]: Make sense.

File: chain/app/db_app.go [hunk 4]
```diff
@@ -244,45 +250,27 @@ func (app *dbBandApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBl
...
-			err := app.dbBand.ResolveRequest(requestID, resolveStatus, rawResult)
-			if err != nil {
-				panic(err)
-			}
-		}
+		app.dbBand.HandleEndblockEvent(event)
 	}
 
+	// Update bonding validator status
+	app.StakingKeeper.IterateValidators(app.DeliverContext,
```
Comment (sorawit) [inline chain/app/db_app.go]: If we are looping like this and already fetch all the data, then why don't we use this loop to update other info as well? Tokens, Jails, DelegatorShares, etc can be updated here. and we will be very sure that we won't mess up. Would that be better?

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go (hunk 4 above)
Comment (sorawit) [reply chain/app/db_app.go]: Now that I think about it. Let's just remove this status thing. It looks quite inefficient and I don't think we need to show it for now. We can certainly revisit later.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by sorawit — DISMISSED** [review]

**Comment by sorawit** [comment]
@taobun Please resolve conflict due to `Pubkey` becoming `string`.

**Review by sorawit — APPROVED** [review]

**Review by tansawit — APPROVED** [review]

**Pull request merged**